From 99abc9c66f605ca907a6291dd1918ec7112db18c Mon Sep 17 00:00:00 2001
Message-ID: <99abc9c66f605ca907a6291dd1918ec7112db18c.1763133105.git.jdenemar@redhat.com>
From: Michal Privoznik <mprivozn@redhat.com>
Date: Tue, 30 Sep 2025 10:27:27 +0200
Subject: [PATCH] conf: More hyperv related members into a single struct
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

So far, we have an array of integers (hyperv_features), an uint
(hyperv_spinlocks), a string (hyperv_vendor_id) and some tristate
switches scattered across virDomainDef. Soon, new knobs will be
introduced and keeping the current state would only worsen
readability.

Introduce virDomainHypervFeatures struct to place hyperv related
features there.

Signed-off-by: Michal Privoznik <mprivozn@redhat.com>
Reviewed-by: Ján Tomko <jtomko@redhat.com>
(cherry picked from commit 47271c204e5471f1cbd22cf34c0166455affc095)

Conflicts:
- src/libxl/libxl_conf.c: Original commit changed this file, but
  because 064682ab330535bb1559045b5900398ae7a8d91d is not
  backported, then the change doesn't make sense. The libxl
  driver is not enabled in RHEL anyway.

Resolves: https://issues.redhat.com/browse/RHEL-122930
Signed-off-by: Michal Privoznik <mprivozn@redhat.com>
---
 src/conf/domain_conf.c   | 77 ++++++++++++++++++++++------------------
 src/conf/domain_conf.h   | 15 ++++----
 src/conf/virconftypes.h  |  2 ++
 src/qemu/qemu_command.c  | 16 ++++-----
 src/qemu/qemu_process.c  |  8 ++---
 src/qemu/qemu_validate.c | 12 +++----
 6 files changed, 71 insertions(+), 59 deletions(-)

diff --git a/src/conf/domain_conf.c b/src/conf/domain_conf.c
index f02efaca4e..67f8ef4676 100644
--- a/src/conf/domain_conf.c
+++ b/src/conf/domain_conf.c
@@ -3949,6 +3949,13 @@ virDomainOSDefClear(virDomainOSDef *os)
 }
 
 
+static void
+virDomainHypervFeaturesClear(virDomainHypervFeatures *hv)
+{
+    g_free(hv->vendor_id);
+}
+
+
 void virDomainDefFree(virDomainDef *def)
 {
     size_t i;
@@ -4076,8 +4083,8 @@ void virDomainDefFree(virDomainDef *def)
     g_free(def->emulator);
     g_free(def->description);
     g_free(def->title);
+    virDomainHypervFeaturesClear(&def->hyperv);
     g_free(def->kvm_features);
-    g_free(def->hyperv_vendor_id);
     g_free(def->tcg_features);
 
     virBlkioDeviceArrayClear(def->blkio.devices,
@@ -16756,7 +16763,7 @@ virDomainFeaturesHyperVDefParse(virDomainDef *def,
                                      &value) < 0)
             return -1;
 
-        def->hyperv_features[feature] = value;
+        def->hyperv.features[feature] = value;
 
         switch ((virDomainHyperv) feature) {
         case VIR_DOMAIN_HYPERV_RELAXED:
@@ -16782,11 +16789,11 @@ virDomainFeaturesHyperVDefParse(virDomainDef *def,
             while (child) {
                 if (STREQ((const char *)child->name, "direct")) {
                     if (virXMLPropTristateSwitch(child, "state", VIR_XML_PROP_REQUIRED,
-                                                 &def->hyperv_tlbflush_direct) < 0)
+                                                 &def->hyperv.tlbflush_direct) < 0)
                         return -1;
                 } else if (STREQ((const char *)child->name, "extended")) {
                     if (virXMLPropTristateSwitch(child, "state", VIR_XML_PROP_REQUIRED,
-                                                 &def->hyperv_tlbflush_extended) < 0)
+                                                 &def->hyperv.tlbflush_extended) < 0)
                         return -1;
                 } else {
                     virReportError(VIR_ERR_CONFIG_UNSUPPORTED,
@@ -16813,7 +16820,7 @@ virDomainFeaturesHyperVDefParse(virDomainDef *def,
                 }
 
                 if (virXMLPropTristateSwitch(child, "state", VIR_XML_PROP_REQUIRED,
-                                             &def->hyperv_stimer_direct) < 0)
+                                             &def->hyperv.stimer_direct) < 0)
                     return -1;
 
                 child = xmlNextElementSibling(child);
@@ -16825,10 +16832,10 @@ virDomainFeaturesHyperVDefParse(virDomainDef *def,
                 break;
 
             if (virXMLPropUInt(node, "retries", 0, VIR_XML_PROP_REQUIRED,
-                               &def->hyperv_spinlocks) < 0)
+                               &def->hyperv.spinlocks) < 0)
                 return -1;
 
-            if (def->hyperv_spinlocks < 0xFFF) {
+            if (def->hyperv.spinlocks < 0xFFF) {
                 virReportError(VIR_ERR_XML_ERROR, "%s",
                                _("HyperV spinlock retry count must be at least 4095"));
                 return -1;
@@ -16839,15 +16846,15 @@ virDomainFeaturesHyperVDefParse(virDomainDef *def,
             if (value != VIR_TRISTATE_SWITCH_ON)
                 break;
 
-            g_clear_pointer(&def->hyperv_vendor_id, g_free);
+            g_clear_pointer(&def->hyperv.vendor_id, g_free);
 
-            if (!(def->hyperv_vendor_id = virXMLPropString(node, "value"))) {
+            if (!(def->hyperv.vendor_id = virXMLPropString(node, "value"))) {
                 virReportError(VIR_ERR_XML_ERROR, "%s",
                                _("missing 'value' attribute for HyperV feature 'vendor_id'"));
                 return -1;
             }
 
-            if (!STRLIM(def->hyperv_vendor_id, VIR_DOMAIN_HYPERV_VENDOR_ID_MAX)) {
+            if (!STRLIM(def->hyperv.vendor_id, VIR_DOMAIN_HYPERV_VENDOR_ID_MAX)) {
                 virReportError(VIR_ERR_XML_ERROR,
                                _("HyperV vendor_id value must not be more than %1$d characters."),
                                VIR_DOMAIN_HYPERV_VENDOR_ID_MAX);
@@ -16855,7 +16862,7 @@ virDomainFeaturesHyperVDefParse(virDomainDef *def,
             }
 
             /* ensure that the string can be passed to qemu */
-            if (strchr(def->hyperv_vendor_id, ',')) {
+            if (strchr(def->hyperv.vendor_id, ',')) {
                 virReportError(VIR_ERR_XML_ERROR, "%s",
                                _("HyperV vendor_id value is invalid"));
                 return -1;
@@ -21296,12 +21303,12 @@ virDomainDefFeaturesCheckABIStability(virDomainDef *src,
             case VIR_DOMAIN_HYPERV_AVIC:
             case VIR_DOMAIN_HYPERV_EMSR_BITMAP:
             case VIR_DOMAIN_HYPERV_XMM_INPUT:
-                if (src->hyperv_features[i] != dst->hyperv_features[i]) {
+                if (src->hyperv.features[i] != dst->hyperv.features[i]) {
                     virReportError(VIR_ERR_CONFIG_UNSUPPORTED,
                                    _("State of HyperV enlightenment feature '%1$s' differs: source: '%2$s', destination: '%3$s'"),
                                    virDomainHypervTypeToString(i),
-                                   virTristateSwitchTypeToString(src->hyperv_features[i]),
-                                   virTristateSwitchTypeToString(dst->hyperv_features[i]));
+                                   virTristateSwitchTypeToString(src->hyperv.features[i]),
+                                   virTristateSwitchTypeToString(dst->hyperv.features[i]));
                     return false;
                 }
 
@@ -21309,21 +21316,21 @@ virDomainDefFeaturesCheckABIStability(virDomainDef *src,
 
             case VIR_DOMAIN_HYPERV_SPINLOCKS:
                 /* spinlock count matters! */
-                if (src->hyperv_spinlocks != dst->hyperv_spinlocks) {
+                if (src->hyperv.spinlocks != dst->hyperv.spinlocks) {
                     virReportError(VIR_ERR_CONFIG_UNSUPPORTED,
                                    _("HyperV spinlock retry count differs: source: '%1$u', destination: '%2$u'"),
-                                   src->hyperv_spinlocks,
-                                   dst->hyperv_spinlocks);
+                                   src->hyperv.spinlocks,
+                                   dst->hyperv.spinlocks);
                     return false;
                 }
                 break;
 
             case VIR_DOMAIN_HYPERV_VENDOR_ID:
-                if (STRNEQ_NULLABLE(src->hyperv_vendor_id, dst->hyperv_vendor_id)) {
+                if (STRNEQ_NULLABLE(src->hyperv.vendor_id, dst->hyperv.vendor_id)) {
                     virReportError(VIR_ERR_CONFIG_UNSUPPORTED,
                                    _("HyperV vendor_id differs: source: '%1$s', destination: '%2$s'"),
-                                   src->hyperv_vendor_id,
-                                   dst->hyperv_vendor_id);
+                                   src->hyperv.vendor_id,
+                                   dst->hyperv.vendor_id);
                     return false;
                 }
                 break;
@@ -21334,12 +21341,12 @@ virDomainDefFeaturesCheckABIStability(virDomainDef *src,
         }
     }
 
-    if (src->hyperv_features[VIR_DOMAIN_HYPERV_STIMER] == VIR_TRISTATE_SWITCH_ON) {
-        if (src->hyperv_stimer_direct != dst->hyperv_stimer_direct) {
+    if (src->hyperv.features[VIR_DOMAIN_HYPERV_STIMER] == VIR_TRISTATE_SWITCH_ON) {
+        if (src->hyperv.stimer_direct != dst->hyperv.stimer_direct) {
             virReportError(VIR_ERR_CONFIG_UNSUPPORTED,
                            _("State of HyperV stimer direct feature differs: source: '%1$s', destination: '%2$s'"),
-                           virTristateSwitchTypeToString(src->hyperv_stimer_direct),
-                           virTristateSwitchTypeToString(dst->hyperv_stimer_direct));
+                           virTristateSwitchTypeToString(src->hyperv.stimer_direct),
+                           virTristateSwitchTypeToString(dst->hyperv.stimer_direct));
             return false;
         }
     }
@@ -28040,11 +28047,11 @@ virDomainFeaturesHyperVDefFormat(virBuffer *buf,
         g_auto(virBuffer) hypervAttrBuf = VIR_BUFFER_INITIALIZER;
         g_auto(virBuffer) hypervChildBuf = VIR_BUFFER_INIT_CHILD(&childBuf);
 
-        if (def->hyperv_features[j] == VIR_TRISTATE_SWITCH_ABSENT)
+        if (def->hyperv.features[j] == VIR_TRISTATE_SWITCH_ABSENT)
             continue;
 
         virBufferAsprintf(&hypervAttrBuf, " state='%s'",
-                          virTristateSwitchTypeToString(def->hyperv_features[j]));
+                          virTristateSwitchTypeToString(def->hyperv.features[j]));
 
         switch ((virDomainHyperv) j) {
         case VIR_DOMAIN_HYPERV_RELAXED:
@@ -28063,34 +28070,34 @@ virDomainFeaturesHyperVDefFormat(virBuffer *buf,
             break;
 
         case VIR_DOMAIN_HYPERV_SPINLOCKS:
-            if (def->hyperv_features[j] == VIR_TRISTATE_SWITCH_ON) {
+            if (def->hyperv.features[j] == VIR_TRISTATE_SWITCH_ON) {
                 virBufferAsprintf(&hypervAttrBuf,
-                                  " retries='%d'", def->hyperv_spinlocks);
+                                  " retries='%d'", def->hyperv.spinlocks);
             }
             break;
 
         case VIR_DOMAIN_HYPERV_STIMER:
-            if (def->hyperv_features[j] == VIR_TRISTATE_SWITCH_ON &&
-                def->hyperv_stimer_direct == VIR_TRISTATE_SWITCH_ON) {
+            if (def->hyperv.features[j] == VIR_TRISTATE_SWITCH_ON &&
+                def->hyperv.stimer_direct == VIR_TRISTATE_SWITCH_ON) {
                 virBufferAddLit(&hypervChildBuf, "<direct state='on'/>\n");
             }
 
             break;
 
         case VIR_DOMAIN_HYPERV_VENDOR_ID:
-            if (def->hyperv_features[j] == VIR_TRISTATE_SWITCH_ON) {
+            if (def->hyperv.features[j] == VIR_TRISTATE_SWITCH_ON) {
                 virBufferEscapeString(&hypervAttrBuf, " value='%s'",
-                                      def->hyperv_vendor_id);
+                                      def->hyperv.vendor_id);
             }
             break;
 
         case VIR_DOMAIN_HYPERV_TLBFLUSH:
-            if (def->hyperv_features[j] != VIR_TRISTATE_SWITCH_ON)
+            if (def->hyperv.features[j] != VIR_TRISTATE_SWITCH_ON)
                 break;
 
-            if (def->hyperv_tlbflush_direct == VIR_TRISTATE_SWITCH_ON)
+            if (def->hyperv.tlbflush_direct == VIR_TRISTATE_SWITCH_ON)
                 virBufferAddLit(&hypervChildBuf, "<direct state='on'/>\n");
-            if (def->hyperv_tlbflush_extended == VIR_TRISTATE_SWITCH_ON)
+            if (def->hyperv.tlbflush_extended == VIR_TRISTATE_SWITCH_ON)
                 virBufferAddLit(&hypervChildBuf, "<extended state='on'/>\n");
             break;
 
diff --git a/src/conf/domain_conf.h b/src/conf/domain_conf.h
index 24a1089552..643deefec2 100644
--- a/src/conf/domain_conf.h
+++ b/src/conf/domain_conf.h
@@ -3080,6 +3080,14 @@ struct _virDomainPstoreDef {
     virDomainDeviceInfo info;
 };
 
+struct _virDomainHypervFeatures {
+    int features[VIR_DOMAIN_HYPERV_LAST];
+    unsigned int spinlocks;
+    virTristateSwitch stimer_direct;
+    virTristateSwitch tlbflush_direct;
+    virTristateSwitch tlbflush_extended;
+    char *vendor_id;
+};
 
 #define SCSI_SUPER_WIDE_BUS_MAX_CONT_UNIT 64
 #define SCSI_WIDE_BUS_MAX_CONT_UNIT 16
@@ -3147,19 +3155,14 @@ struct _virDomainDef {
      * See virDomainDefFeaturesCheckABIStability() for details. */
     int features[VIR_DOMAIN_FEATURE_LAST];
     int caps_features[VIR_DOMAIN_PROCES_CAPS_FEATURE_LAST];
-    int hyperv_features[VIR_DOMAIN_HYPERV_LAST];
+    virDomainHypervFeatures hyperv;
     virDomainFeatureKVM *kvm_features;
     int msrs_features[VIR_DOMAIN_MSRS_LAST];
     int xen_features[VIR_DOMAIN_XEN_LAST];
     virDomainXenPassthroughMode xen_passthrough_mode;
-    unsigned int hyperv_spinlocks;
-    virTristateSwitch hyperv_stimer_direct;
-    virTristateSwitch hyperv_tlbflush_direct;
-    virTristateSwitch hyperv_tlbflush_extended;
     virGICVersion gic_version;
     virDomainHPTResizing hpt_resizing;
     unsigned long long hpt_maxpagesize; /* Stored in KiB */
-    char *hyperv_vendor_id;
     virTristateSwitch apic_eoi;
     virDomainFeatureTCG *tcg_features;
 
diff --git a/src/conf/virconftypes.h b/src/conf/virconftypes.h
index d46da4bdda..e93d260d37 100644
--- a/src/conf/virconftypes.h
+++ b/src/conf/virconftypes.h
@@ -138,6 +138,8 @@ typedef struct _virDomainHubDef virDomainHubDef;
 
 typedef struct _virDomainHugePage virDomainHugePage;
 
+typedef struct _virDomainHypervFeatures virDomainHypervFeatures;
+
 typedef struct _virDomainIOMMUDef virDomainIOMMUDef;
 
 typedef struct _virDomainIOThreadIDDef virDomainIOThreadIDDef;
diff --git a/src/qemu/qemu_command.c b/src/qemu/qemu_command.c
index 13e23c7ee3..3f290931a1 100644
--- a/src/qemu/qemu_command.c
+++ b/src/qemu/qemu_command.c
@@ -6320,7 +6320,7 @@ qemuBuildCpuHypervCommandLine(virBuffer *buf,
         case VIR_DOMAIN_HYPERV_AVIC:
         case VIR_DOMAIN_HYPERV_EMSR_BITMAP:
         case VIR_DOMAIN_HYPERV_XMM_INPUT:
-            if (def->hyperv_features[i] == VIR_TRISTATE_SWITCH_ON) {
+            if (def->hyperv.features[i] == VIR_TRISTATE_SWITCH_ON) {
                 const char *name = virDomainHypervTypeToString(i);
                 g_autofree char *full_name = g_strdup_printf("hv-%s", name);
                 const char *qemu_name = virQEMUCapsCPUFeatureToQEMU(def->os.arch,
@@ -6328,27 +6328,27 @@ qemuBuildCpuHypervCommandLine(virBuffer *buf,
                 virBufferAsprintf(buf, ",%s=on", qemu_name);
             }
             if ((i == VIR_DOMAIN_HYPERV_STIMER) &&
-                (def->hyperv_stimer_direct == VIR_TRISTATE_SWITCH_ON))
+                (def->hyperv.stimer_direct == VIR_TRISTATE_SWITCH_ON))
                 virBufferAsprintf(buf, ",%s=on", VIR_CPU_x86_HV_STIMER_DIRECT);
             if (i == VIR_DOMAIN_HYPERV_TLBFLUSH) {
-                if (def->hyperv_tlbflush_direct == VIR_TRISTATE_SWITCH_ON)
+                if (def->hyperv.tlbflush_direct == VIR_TRISTATE_SWITCH_ON)
                     virBufferAsprintf(buf, ",%s=on", VIR_CPU_x86_HV_TLBFLUSH_DIRECT);
-                if (def->hyperv_tlbflush_extended == VIR_TRISTATE_SWITCH_ON)
+                if (def->hyperv.tlbflush_extended == VIR_TRISTATE_SWITCH_ON)
                     virBufferAsprintf(buf, ",%s=on", VIR_CPU_x86_HV_TLBFLUSH_EXT);
             }
             break;
 
         case VIR_DOMAIN_HYPERV_SPINLOCKS:
-            if (def->hyperv_features[i] == VIR_TRISTATE_SWITCH_ON)
+            if (def->hyperv.features[i] == VIR_TRISTATE_SWITCH_ON)
                 virBufferAsprintf(buf, ",%s=0x%x",
                                   VIR_CPU_x86_HV_SPINLOCKS,
-                                  def->hyperv_spinlocks);
+                                  def->hyperv.spinlocks);
             break;
 
         case VIR_DOMAIN_HYPERV_VENDOR_ID:
-            if (def->hyperv_features[i] == VIR_TRISTATE_SWITCH_ON)
+            if (def->hyperv.features[i] == VIR_TRISTATE_SWITCH_ON)
                 virBufferAsprintf(buf, ",hv-vendor-id=%s",
-                                  def->hyperv_vendor_id);
+                                  def->hyperv.vendor_id);
             break;
 
         case VIR_DOMAIN_HYPERV_LAST:
diff --git a/src/qemu/qemu_process.c b/src/qemu/qemu_process.c
index 7b6c02bc27..798dabaf0a 100644
--- a/src/qemu/qemu_process.c
+++ b/src/qemu/qemu_process.c
@@ -4381,7 +4381,7 @@ qemuProcessVerifyHypervFeatures(virDomainDef *def,
             i == VIR_DOMAIN_HYPERV_SPINLOCKS)
             continue;
 
-        if (def->hyperv_features[i] != VIR_TRISTATE_SWITCH_ON)
+        if (def->hyperv.features[i] != VIR_TRISTATE_SWITCH_ON)
             continue;
 
         cpuFeature = g_strdup_printf("hv-%s", virDomainHypervTypeToString(i));
@@ -4392,7 +4392,7 @@ qemuProcessVerifyHypervFeatures(virDomainDef *def,
             return -1;
         } else if (rc == 1) {
             if (i == VIR_DOMAIN_HYPERV_STIMER) {
-                if (def->hyperv_stimer_direct != VIR_TRISTATE_SWITCH_ON)
+                if (def->hyperv.stimer_direct != VIR_TRISTATE_SWITCH_ON)
                     continue;
 
                 rc = virCPUDataCheckFeature(cpu, VIR_CPU_x86_HV_STIMER_DIRECT);
@@ -4407,7 +4407,7 @@ qemuProcessVerifyHypervFeatures(virDomainDef *def,
                 return -1;
             }
             if (i == VIR_DOMAIN_HYPERV_TLBFLUSH) {
-                if (def->hyperv_tlbflush_direct == VIR_TRISTATE_SWITCH_ON) {
+                if (def->hyperv.tlbflush_direct == VIR_TRISTATE_SWITCH_ON) {
                     rc = virCPUDataCheckFeature(cpu, VIR_CPU_x86_HV_TLBFLUSH_DIRECT);
                     if (rc < 0)
                         return -1;
@@ -4418,7 +4418,7 @@ qemuProcessVerifyHypervFeatures(virDomainDef *def,
                         return -1;
                     }
                 }
-                if (def->hyperv_tlbflush_extended == VIR_TRISTATE_SWITCH_ON) {
+                if (def->hyperv.tlbflush_extended == VIR_TRISTATE_SWITCH_ON) {
                     rc = virCPUDataCheckFeature(cpu, VIR_CPU_x86_HV_TLBFLUSH_EXT);
                     if (rc < 0)
                         return -1;
diff --git a/src/qemu/qemu_validate.c b/src/qemu/qemu_validate.c
index 22565d9e36..ed769efc7c 100644
--- a/src/qemu/qemu_validate.c
+++ b/src/qemu/qemu_validate.c
@@ -89,8 +89,8 @@ qemuValidateDomainDefPSeriesFeature(const virDomainDef *def,
 
 
 #define CHECK_HV_FEAT(feat, requires) \
-    if (def->hyperv_features[feat] == VIR_TRISTATE_SWITCH_ON && \
-        def->hyperv_features[requires] != VIR_TRISTATE_SWITCH_ON) { \
+    if (def->hyperv.features[feat] == VIR_TRISTATE_SWITCH_ON && \
+        def->hyperv.features[requires] != VIR_TRISTATE_SWITCH_ON) { \
         virReportError(VIR_ERR_CONFIG_UNSUPPORTED, \
                        _("'%1$s' hyperv feature requires '%2$s' feature"), \
                        virDomainHypervTypeToString(feat), \
@@ -114,7 +114,7 @@ qemuValidateDomainDefHypervFeatures(const virDomainDef *def)
 
     CHECK_HV_FEAT(VIR_DOMAIN_HYPERV_SYNIC, VIR_DOMAIN_HYPERV_VPINDEX);
 
-    if (def->hyperv_features[VIR_DOMAIN_HYPERV_STIMER] == VIR_TRISTATE_SWITCH_ON) {
+    if (def->hyperv.features[VIR_DOMAIN_HYPERV_STIMER] == VIR_TRISTATE_SWITCH_ON) {
         if (!virDomainDefHasTimer(def, VIR_DOMAIN_TIMER_NAME_HYPERVCLOCK)) {
             virReportError(VIR_ERR_CONFIG_UNSUPPORTED,
                            _("'%1$s' hyperv feature requires '%2$s' timer"),
@@ -133,9 +133,9 @@ qemuValidateDomainDefHypervFeatures(const virDomainDef *def)
 
     CHECK_HV_FEAT(VIR_DOMAIN_HYPERV_EVMCS, VIR_DOMAIN_HYPERV_VAPIC);
 
-    if (def->hyperv_features[VIR_DOMAIN_HYPERV_TLBFLUSH] == VIR_TRISTATE_SWITCH_ON &&
-        def->hyperv_tlbflush_direct == VIR_TRISTATE_SWITCH_ON) {
-        if (def->hyperv_features[VIR_DOMAIN_HYPERV_VAPIC] != VIR_TRISTATE_SWITCH_ON) {
+    if (def->hyperv.features[VIR_DOMAIN_HYPERV_TLBFLUSH] == VIR_TRISTATE_SWITCH_ON &&
+        def->hyperv.tlbflush_direct == VIR_TRISTATE_SWITCH_ON) {
+        if (def->hyperv.features[VIR_DOMAIN_HYPERV_VAPIC] != VIR_TRISTATE_SWITCH_ON) {
             virReportError(VIR_ERR_CONFIG_UNSUPPORTED,
                            _("'%1$s' hyperv feature requires '%2$s' feature"),
                            VIR_CPU_x86_HV_TLBFLUSH_DIRECT,
-- 
2.51.1
